core/src/cluster_info.rs

```diff
@@ -1076,22 +1077,60 @@ impl ClusterInfo {
         self.push_message(CrdsValue::new_signed(message, &self.keypair));
     }
 
-    pub fn push_vote(&self, tower_index: usize, vote: Transaction) {
+    pub fn push_vote(&self, tower: &[Slot], vote: Transaction) {
+        debug_assert!(tower.iter().tuple_windows().all(|(a, b)| a < b));
         let now = timestamp();
-        let vote = Vote::new(&self.id(), vote, now);
-        let vote_ix = {
-            let r_gossip =
+        // Find a crds vote which is evicted from the tower, and recycle its
+        // vote-index. This can be either an old vote which is popped off the
+        // deque, or recent vote which has expired before getting enough
+        // confirmations.
+        // If all votes are still in the tower, add a new vote-index. If more
+        // than one vote is evicted, the oldest one by wallclock is returned in
+        // order to allow more recent votes more time to propagate through
+        // gossip.
+        // TODO: When there are more than one vote evicted from the tower, only
+        // one crds vote is overwritten here. Decide what to do with the rest.
+        let mut num_crds_votes = 0;
+        let self_pubkey = self.id();
+        // Returns true if the tower does not contain the vote.slot.
+        let should_evict_vote = |vote: &Vote| -> bool {
+            match vote.slot() {
+                Some(slot) => !tower.contains(&slot),
+                None => {
+                    error!("crds vote with no slots!");
+                    true
+                }
+            }
+        };
+        let vote_index = {
+            let gossip =
                 self.time_gossip_read_lock("gossip_read_push_vote", &self.stats.push_vote_read);
-            let current_votes: Vec<_> = (0..crds_value::MAX_VOTES)
-                .filter_map(|ix| r_gossip.crds.lookup(&CrdsValueLabel::Vote(ix, self.id())))
-                .collect();
-            CrdsValue::compute_vote_index(tower_index, current_votes)
+            (0..MAX_LOCKOUT_HISTORY as u8)
+                .filter_map(|ix| {
+                    let vote = CrdsValueLabel::Vote(ix, self_pubkey);
+                    let vote = gossip.crds.lookup(&vote)?;
+                    num_crds_votes += 1;
+                    match &vote.data {
+                        CrdsData::Vote(_, vote) if should_evict_vote(vote) => {
+                            Some((vote.wallclock, ix))
+                        }
+                        CrdsData::Vote(_, _) => None,
+                        _ => panic!("this should not happen!"),
+                    }
+                })
+                .min() // Boot the oldest evicted vote by wallclock.
```

Comment:
Can we add a test that:
1)  When we have the MAX_LOCKOUT_HISTORY votes, pushing a new vote actually pops off the first one (confirm behavior of the current bug is gone).
2) When multiple votes are up for eviction, the oldest one is evicted.

Reply:
added more test coverage

> When we have the MAX_LOCKOUT_HISTORY votes, pushing a new vote actually pops off the first one

It will pop off whatever is not in the `tower: &[Slot]` argument. It can be the first or it can be the last.